File: src/2020/AdventOfCode2020/Day02.cs
using AdventOfCodeLib;

namespace AdventOfCode2020;

public sealed class Day02 : IAocDay<int>
{
    public static int Part1(AocInput input) => input.Lines
        .Select(ParseLine)
        .Select(CreatePasswordPolicy)
        .Count(IsPasswordPolicyValid);

    public static int Part2(AocInput input) => input.Lines
        .Select(ParseLine)
        .Select(CreatePasswordPolicy)
        .Count(IsNewPasswordPolicyValid);
    
    private static IEnumerable<string> ParseLine(string line) => 
        line.Split(Separators, StringSplitOptions.TrimEntries);

    private static PasswordPolicy CreatePasswordPolicy(IEnumerable<string> segments)
    {
        var parts = segments.ToArray();
        return new PasswordPolicy(int.Parse(parts[0]), int.Parse(parts[1]), char.Parse(parts[2]), parts[3]);
    }

    private static bool IsPasswordPolicyValid(PasswordPolicy passwordPolicy)
    {
        var countLetter = passwordPolicy.Password.Count(c => c == passwordPolicy.Letter);
        return passwordPolicy.Min <= countLetter && countLetter <= passwordPolicy.Max;
    }

    private static bool IsNewPasswordPolicyValid(PasswordPolicy passwordPolicy)
    {
        var firstElement = passwordPolicy.Password[passwordPolicy.Min - 1];
        var secondElement = passwordPolicy.Password[passwordPolicy.Max - 1];
        return new[] { firstElement, secondElement }
            .Count(m => m == passwordPolicy.Letter) is 1;
    }

    private static readonly string[] Separators = ["-", " ", ": "];

    private sealed record PasswordPolicy(int Min, int Max, char Letter, string Password);
}

File: python/y2020/day02.py
from os import pipe
import re
from aoc.core import compose, pipe, solve, AocInput
from dataclasses import astuple, dataclass


@dataclass(frozen=True, slots=True)
class PasswordPolicy:
    min: int
    max: int
    letter: str
    password: str

    def __iter__(self):
        return iter(astuple(self))

    @staticmethod
    def create(values: tuple[str]) -> "PasswordPolicy":
        return PasswordPolicy(int(values[0]), int(values[1]), values[2], values[3])


def parseLine(line: str) -> tuple[str]:
    parts = re.split("-| |: ", line)
    return tuple([s.strip() for s in parts])


def isPasswordPolicyValid(posswordPolicy: PasswordPolicy) -> bool:
    min, max, letter, password = posswordPolicy
    letters = len([c for c in list(password) if c == letter])
    return min <= letters <= max


def isNewPasswordPolicyValid(posswordPolicy: PasswordPolicy) -> bool:
    min, max, letter, password = posswordPolicy
    pw1 = password[min - 1]
    pw2 = password[max - 1]
    return len([c for c in [pw1, pw2] if c == letter]) == 1


def part1(input: AocInput) -> int:
    return sum(map(compose(isNewPasswordPolicyValid, PasswordPolicy.create, parseLine), input.lines))


def part2(input: AocInput) -> int:
    return sum(map(pipe(parseLine, PasswordPolicy.create, isNewPasswordPolicyValid), input.lines))


def main() -> None:
    YEAR = "y2020"
    DAY = "Day02"
    print(f"Running year {YEAR.replace('y', '')} day {DAY.replace('Day', '')}")
    solve(YEAR, DAY, part1)
    solve(YEAR, DAY, part2)
